### PR TITLE
Use ecma-collection in JSON.

### PR DESCRIPTION
**Pull request by rtakacs**
The simple list data structure has replaced with ecma-collection in the JSON object.

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp [hunk 1]
```diff
@@ -26,173 +26,68 @@
...
-    if (*current_p == element_p)
-    {
-      return true;
-    }
-  }
+  ecma_object_t *obj_p = ecma_get_object_from_value (object_value);
 
-  return false;
-} /* list_has_element */
+  ecma_collection_iterator_t iterator;
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp]: Maybe, the function should be named `ecma_has_object_value_in_collection`.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 2]
```diff
@@ -1563,31 +1553,30 @@ ecma_builtin_json_object (ecma_object_t *obj_p, /**< the object*/
...
   }
 
   /* 7. */
-  list_ctx_t partial;
-  {
-    partial.block_start_p = NULL;
-    partial.block_end_p = NULL;
-    partial.current_p = NULL;
-  }
+  ecma_collection_header_t *partial = ecma_new_values_collection (NULL, 0, true);
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: `_p` suffix is missing.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 3]
```diff
@@ -1541,21 +1534,18 @@ ecma_builtin_json_object (ecma_object_t *obj_p, /**< the object*/
   /* 4. */
   context_p->indent_str_p = ecma_concat_ecma_strings (context_p->indent_str_p, context_p->gap_str_p);
 
-  list_ctx_t property_keys;
-  {
-    property_keys.block_start_p = NULL;
-    property_keys.block_end_p = NULL;
-    property_keys.current_p = NULL;
-  }
+  ecma_collection_header_t *property_keys;
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: `_p`

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h [hunk 4]
```diff
@@ -130,45 +130,36 @@ extern ecma_completion_value_t ecma_date_get_primitive_value (ecma_value_t this_
...
-  void **block_end_p;
-  void **current_p;
-} list_ctx_t;
-
-typedef struct
-{
-  list_ctx_t property_list;
-  list_ctx_t occurence_stack;
+  ecma_collection_header_t *property_list;
+  ecma_collection_header_t *occurence_stack;
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h]: Comments to structure and its elements are missing.

**Comment by rtakacs** [comment]
I've updated the pull request.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h [hunk 5]
```diff
@@ -130,45 +130,48 @@ extern ecma_completion_value_t ecma_date_get_primitive_value (ecma_value_t this_
...
 
+  /** The actual indentation text. */
   ecma_string_t *indent_str_p;
+
+  /** The indentation text. */
   ecma_string_t *gap_str_p;
+
+  /** The replacer function. */
   ecma_object_t *replacer_function_p;
 } stringify_context_t;
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h]: Could you, please, also add `ecma_json_` or `ecma_` prefix to the structure name?

**Comment by ruben-ayrapetyan** [comment]
Looks good to me

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h (hunk 5 above)
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h]: Of course. Done.

**Comment by dbatyai** [comment]
lgtm

**Comment by zherczeg** [comment]
LGTM

**Comment by rtakacs** [comment]
I think we should close this issue.

**Pull request merged**